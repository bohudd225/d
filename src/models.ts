export type SocialNetworksClientIds = {
  facebook?: string,
  google?: string,
  linkedin?: string
}

export type FormOptions = {
  fields: {
    firstName?: string,
    lastName?: string,
    dateOfBirth?: string
  }
}

export type Options = {
  socialNetworksClientIds: SocialNetworksClientIds,
  form: FormOptions
}
