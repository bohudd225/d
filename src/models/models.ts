import * as t from 'tcomb';
import FbGraphApi from './facebook-api';
import { BaseProperties as CustomerBase, Social as CustomerSocial, Contacts as CustomerContacts, ContactHubSDKBrowser } from './contacthub-sdk-browser';

export type SocialNetworksClientIds = {
  facebook?: string,
  google?: string,
  linkedin?: string
}

export type SocialNetwork = keyof SocialNetworksClientIds

export type Fields = {
  firstName?: string,
  lastName?: string,
  dateOfBirth?: string,
  email?: string,
  gender?: string
}

export type Icons = {
  container: string
}

export type AutofillOptions = {
  icons?: Icons,
  fields: Fields
}

export type Scope = 'likes' | 'education_history' | 'work_history'

export type Options = {
  socialNetworks: {
    clientIds: SocialNetworksClientIds,
    scopes?: Scope[]
  },
  contacthub: ContactHubSDKBrowser,
  autofillOptions: AutofillOptions
}

// tcomb version of Options to validate at run-time
export const Options = t.interface({
  socialNetworks: t.interface({
    clientIds: t.interface({
      facebook: t.maybe(t.String),
      google: t.maybe(t.String),
      linkedin: t.maybe(t.String)
    }, { strict: true }),
    scopes: t.maybe(t.list(t.enums.of(['likes', 'education_history', 'work_history'])))
  }, { strict: true }),
  contacthub: t.Function,
  autofillOptions: t.interface({
    icons: t.maybe(t.interface({
      container: t.String
    })),
    fields: t.interface({
      firstName: t.maybe(t.String),
      lastName: t.maybe(t.String),
      dateOfBirth: t.maybe(t.String),
      email: t.maybe(t.String),
      gender: t.maybe(t.String)
    }, { strict: true })
  }, { strict: true })
}, { strict: true, name: '"ContacthubConnectSocial"' });

export type FacebookScope = 'email' | 'user_birthday' | 'user_likes' | 'user_education_history' | 'user_work_history'
export type GoogleScope = 'email' | 'birthday'

export type User = {
  base: CustomerBase & {
    contacts: CustomerContacts,
    socialProfile: CustomerSocial
  }
}

export interface HelloUser {
  first_name: string,
  last_name: string,
  name: string,
  picture: string,
  thumbnail: string,
  email: string
}

export type FacebookFields = 'id' | 'first_name' | 'last_name' | 'gender' | 'birthday' | 'link' | 'email' | 'education' | 'work';
export type FacebookUser = Pick<FbGraphApi.FbUser, FacebookFields> & HelloUser
export type FacebookLike = {
  created_time: string,
  id: string,
  name: string,
  pictures?: any[]
}

// copied from https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/gapi.plus/index.d.ts
// see https://developers.google.com/+/web/api/rest/latest/people for official docs from Google
export interface GoogleUser extends HelloUser {
  kind: 'plus#person';
  etag: string;
  nickname: string;
  occupation: string;
  skills: string;
  birthday: string;
  gender: string;
  emails: {
    value: string;
    type: string;
  }[];
  urls: {
    value: string;
    type: string;
    label: string;
  }[];
  objectType: string;
  id: string;
  displayName: string;
  /* overwritten by hellojs (see HelloUser['name'] type) */
  // name: {
  //   formatted: string;
  //   familyName: string;
  //   givenName: string;
  //   middleName: string;
  //   honorificPrefix: string;
  //   honorificSuffix: string;
  // };
  tagline: string;
  braggingRights: string;
  aboutMe: string;
  relationshipStatus: string;
  url: string;
  image: {
    url: string;
  };
  organizations: {
    name?: string;
    department?: string;
    title?: string;
    type: 'school' | 'work';
    startDate?: string;
    endDate?: string;
    location?: string;
    description?: string;
    primary?: boolean;
  }[];
  placesLived: {
    value: string;
    primary: boolean;
  }[];
  isPlusUser: boolean;
  language: string;
  ageRange: {
    min: number;
    max: number;
  };
  plusOneCount: number;
  circledByCount: number;
  verified: boolean;
  cover: {
    layout: string;
    coverPhoto: {
      url: string;
      height: number;
      width: number;
    };
    coverInfo: {
      topImageOffset: number;
      leftImageOffset: number;
    }
  };
  domain: string;
}

/*
  cherry-picked the keys we use from docs here: https://developer.linkedin.com/docs/fields/basic-profile
  note: linkedin sdk (apparently) converts each key to camelCase
*/

export type LinkedInUser = {
  id: string,
  firstName: string,
  lastName: string,
  formattedName: string,
  pictureUrl: string,
  publicProfileUrl: string,
  emailAddress: string,
  birthDate?: {
    day: number,
    month: number,
    year: number
  },
  educations?: {
    values: {
      id: number,
      activities?: never,
      degreeName?: never,
      endDate?: { year: number, month: number },
      startDate?: { year: number, month: number },
      fieldsOfStudy?: never,
      grade?: never,
      notes?: never,
      program?: never,
      richMediaAssociations?: never,
      school?: {
        name: string
      },
      schoolName?: never,
      isCurrent?: boolean
    }[]
  },
  positions?: {
    values: {
      id: number,
      company?: {
        id: number,
        name: string,
        industry: string
      },
      companyName?: never,
      description?: never,
      endDate?: { year: number, month: number },
      startDate?: { year: number, month: number },
      location?: never,
      locationName?: never,
      richMediaAssociations?: never,
      title?: string,
      isCurrent?: boolean
    }[]
  }

}
