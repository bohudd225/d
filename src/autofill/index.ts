import FormFiller from './formFiller';
import addSocialIconsToForm from './addSocialIconsToForm';
export { FormFiller, addSocialIconsToForm };
