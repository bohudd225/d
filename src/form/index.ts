import { FormOptions } from '../models';

export default function(formOptions: FormOptions) {
  console.log(formOptions);
}
