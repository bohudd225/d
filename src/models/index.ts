export { CustomerData, ContactHubSDKBrowser } from './contacthub-sdk-browser';
export * from './models';
