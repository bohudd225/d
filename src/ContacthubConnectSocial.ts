import { Promise } from 'es6-promise';
import { FormFiller, addSocialIconsToForm } from './autofill';
import SocialNetworks from './socialNetworks';
import { Options, ContactHubSDKBrowser, User } from './models';

export default class ContacthubConnectSocial {

  socialNetworks: SocialNetworks
  contacthub: ContactHubSDKBrowser
  formFiller: FormFiller

  constructor(options: Options) {
    if (typeof options === 'undefined') {
      throw new Error('"Social Autofill" must be initialized with some options');
    }

    // fully validate "options" with tcomb
    Options(options as any);

    this.socialNetworks = new SocialNetworks(options.clientIds);
    this.contacthub = options.contacthub;
    this.formFiller = new FormFiller(options.autofillOptions.fields);

    if (options.autofillOptions.icons) {
      addSocialIconsToForm(options.autofillOptions.icons, this.socialNetworks, this);
    }
  }

  sendUserToContactHub = (user: User) => this.contacthub('customer', user)

  loginWithFacebook(): Promise<User> {
    return this.socialNetworks.loginWithFacebook()
      .then(user => {
        this.sendUserToContactHub(user);
        this.formFiller.fillFormWithUserData(user);
        return user;
      });
  }

  loginWithGoogle(): Promise<User> {
    return this.socialNetworks.loginWithGoogle()
      .then(user => {
        this.sendUserToContactHub(user);
        this.formFiller.fillFormWithUserData(user);
        return user;
      });
  }

  loginWithLinkedIn(): Promise<User> {
    return this.socialNetworks.loginWithLinkedIn()
      .then(user => {
        this.sendUserToContactHub(user);
        this.formFiller.fillFormWithUserData(user);
        return user;
      });
  }
}
