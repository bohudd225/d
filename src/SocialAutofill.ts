import { Promise } from 'es6-promise';
import { FormFiller } from './form';
import SocialNetworks from './socialNetworks';
import { Options, ContactHubSDKBrowser, User } from './models';

export default class SocialAutofill {

  socialNetworks: SocialNetworks
  contacthub: ContactHubSDKBrowser
  formFiller: FormFiller

  constructor(options: Options) {
    if (typeof options === 'undefined') {
      throw new Error('"Social Autofill" must be initialized with some options');
    }

    // fully validate "options" with tcomb
    Options(options as any);

    this.socialNetworks = new SocialNetworks(options.clientIds);
    this.contacthub = options.contacthub;
    this.formFiller = new FormFiller(options.autofillOptions.fields);
  }

  sendUserToContactHub = (user: User) => this.contacthub('customer', user)

  loginWithFacebook(): Promise<void> {
    return this.socialNetworks.loginWithFacebook()
      .then(user => {
        this.sendUserToContactHub(user);
        this.formFiller.fillFormWithUserData(user);
      });
  }

  loginWithGoogle(): Promise<void> {
    return this.socialNetworks.loginWithGoogle()
      .then(user => {
        this.sendUserToContactHub(user);
        this.formFiller.fillFormWithUserData(user);
      });
  }

  loginWithLinkedIn(): Promise<void> {
    return this.socialNetworks.loginWithLinkedIn()
      .then(user => {
        this.sendUserToContactHub(user);
        this.formFiller.fillFormWithUserData(user);
      });
  }
}
