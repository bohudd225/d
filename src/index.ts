import form from './form';
import socialNetworks from './socialNetworks';
import { Options } from './models';

module.exports = function SocialAutofill(options: Options) {
  if (typeof options === 'undefined') {
    throw new Error('"Social Autofill" must be initialized with some options');
  }
  socialNetworks(options.socialNetworksAppIds);
  form(options.form);
}
