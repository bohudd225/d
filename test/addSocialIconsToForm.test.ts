import { User } from '../src/models';
import SocialAutofill from '../src/SocialAutofill';

const validOptions = {
  clientIds: { facebook: '123456', google: '123456', linkedin: '123456' },
  contacthub: () => {},
  autofillOptions: { icons: { container: '#icons-container' }, fields: {} }
}

describe('addSocialIconsToForm function', () => {

  it('add social network icons to icons container element', () => {

    document.body.innerHTML = `
      <form class="ui form">
        <div id="icons-container"></div>
      </form>`;

    new SocialAutofill(validOptions);
    expect(document.body.innerHTML).toMatchSnapshot();
  });

});
