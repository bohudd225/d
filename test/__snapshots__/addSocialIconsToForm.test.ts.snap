// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addSocialIconsToForm function add social network icons to icons container element 1`] = `
"
      <form class=\\"ui form\\">
        <div id=\\"icons-container\\"><svg fill=\\"#3b5998\\" id=\\"social-autofill-facebook-icon\\" width=\\"30px\\" style=\\"cursor:pointer;\\" onclick=\\"__contacthubConnectSocial.loginWithFacebook()\\" aria-labelledby=\\"title\\" role=\\"img\\" viewBox=\\"0 0 24 24\\" xmlns=\\"http://www.w3.org/2000/svg\\"><title>Facebook icon</title><path d=\\"M22.676 0H1.324C.593 0 0 .593 0 1.324v21.352C0 23.408.593 24 1.324 24h11.494v-9.294H9.689v-3.621h3.129V8.41c0-3.099 1.894-4.785 4.659-4.785 1.325 0 2.464.097 2.796.141v3.24h-1.921c-1.5 0-1.792.721-1.792 1.771v2.311h3.584l-.465 3.63H16.56V24h6.115c.733 0 1.325-.592 1.325-1.324V1.324C24 .593 23.408 0 22.676 0\\"></path></svg>
<svg id=\\"social-autofill-google-icon\\" width=\\"30px\\" style=\\"padding: 2.5px; cursor:pointer;\\" onclick=\\"__contacthubConnectSocial.loginWithGoogle()\\" version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 48 48\\" class=\\"abcRioButtonSvg\\"><g><path fill=\\"#EA4335\\" d=\\"M24 9.5c3.54 0 6.71 1.22 9.21 3.6l6.85-6.85C35.9 2.38 30.47 0 24 0 14.62 0 6.51 5.38 2.56 13.22l7.98 6.19C12.43 13.72 17.74 9.5 24 9.5z\\"></path><path fill=\\"#4285F4\\" d=\\"M46.98 24.55c0-1.57-.15-3.09-.38-4.55H24v9.02h12.94c-.58 2.96-2.26 5.48-4.78 7.18l7.73 6c4.51-4.18 7.09-10.36 7.09-17.65z\\"></path><path fill=\\"#FBBC05\\" d=\\"M10.53 28.59c-.48-1.45-.76-2.99-.76-4.59s.27-3.14.76-4.59l-7.98-6.19C.92 16.46 0 20.12 0 24c0 3.88.92 7.54 2.56 10.78l7.97-6.19z\\"></path><path fill=\\"#34A853\\" d=\\"M24 48c6.48 0 11.93-2.13 15.89-5.81l-7.73-6c-2.15 1.45-4.92 2.3-8.16 2.3-6.26 0-11.57-4.22-13.47-9.91l-7.98 6.19C6.51 42.62 14.62 48 24 48z\\"></path><path fill=\\"none\\" d=\\"M0 0h48v48H0z\\"></path></g></svg>
<svg fill=\\"#0077B5\\" id=\\"social-autofill-linkedin-icon\\" width=\\"30px\\" style=\\"cursor:pointer;\\" onclick=\\"__contacthubConnectSocial.loginWithLinkedIn()\\" aria-labelledby=\\"title\\" role=\\"img\\" viewBox=\\"0 0 24 24\\" xmlns=\\"http://www.w3.org/2000/svg\\"><title>LinkedIn icon</title><path d=\\"M20.447 20.452h-3.554v-5.569c0-1.328-.027-3.037-1.852-3.037-1.853 0-2.136 1.445-2.136 2.939v5.667H9.351V9h3.414v1.561h.046c.477-.9 1.637-1.85 3.37-1.85 3.601 0 4.267 2.37 4.267 5.455v6.286zM5.337 7.433c-1.144 0-2.063-.926-2.063-2.065 0-1.138.92-2.063 2.063-2.063 1.14 0 2.064.925 2.064 2.063 0 1.139-.925 2.065-2.064 2.065zm1.782 13.019H3.555V9h3.564v11.452zM22.225 0H1.771C.792 0 0 .774 0 1.729v20.542C0 23.227.792 24 1.771 24h20.451C23.2 24 24 23.227 24 22.271V1.729C24 .774 23.2 0 22.222 0h.003z\\"></path></svg></div>
      </form>"
`;
